### PR TITLE
[Integration][ArgoCD] Iteration over Null Values Breaking ReSync

### PR DESCRIPTION
**Pull request by dennis-bilson-port**
### **User description**
…s` from API

# Description

What
Added null check for application resources from the ArgoCD API

Why
This provides a better null safety check to resolve issues arising from null values being returned for the resources queried.

How

Implemented a check against instances where:
* `applications` is `None`
* `application` has missing `metadata.uid` field

## Type of change

Please leave one option from the following and delete the rest:

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [x] Integration able to create all default resources from scratch
- [x] Resync finishes successfully
- [x] Resync able to create entities
- [x] Resync able to update entities
- [x] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [x] Integration able to create all default resources from scratch
- [x] Resync able to create entities
- [x] Resync able to update entities
- [x] Resync able to detect and delete entities
- [x] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Bug fix

___

### **Description**
- Fixed null value iteration errors in ArgoCD deployment history and Kubernetes resources

- Added comprehensive test coverage for client methods

- Updated environment configuration example

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["ArgoCD API"] --> B["get_deployment_history()"]
  A --> C["get_kubernetes_resource()"]
  B --> D["Null Check Added"]
  C --> E["Null Check Added"]
  D --> F["Safe Iteration"]
  E --> F
  F --> G["Prevent ReSync Failures"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>client.py</strong><dd><code>Add null value protection for API responses</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/argocd/client.py

<ul><li>Added null checks for applications list in <code>get_deployment_history()</code><br> <li> Added null checks for applications list in <code>get_kubernetes_resource()</code><br> <li> Added safe navigation for status fields using <code>.get()</code> method<br> <li> Added application existence validation in list comprehensions</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1923/files#diff-9a9063248e255966830297ae5cbb17c9b2b93bb6d1a7e4760f5b0a4152619ead">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_client.py</strong><dd><code>Add comprehensive client method test coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/argocd/tests/test_client.py

<ul><li>Created comprehensive test suite for ArgocdClient methods<br> <li> Added tests for <code>get_resources()</code>, <code>get_application_by_name()</code>, <br><code>get_deployment_history()</code><br> <li> Added tests for <code>get_kubernetes_resource()</code> and <code>get_managed_resources()</code><br> <li> Included edge case tests for missing data scenarios</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1923/files#diff-5f1ed807123aa649375fce61a8b3487e3a13508bdee48b160c9e3b6b04d3c944">+265/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Miscellaneous</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_sample.py</strong><dd><code>Remove placeholder test</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/argocd/tests/test_sample.py

- Removed placeholder test function

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1923/files#diff-a44bae4e4a6b52850253e021f675bdf8a9ee045757d6f0e63728045ca95a0d31">+0/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>.env.example</strong><dd><code>Update environment configuration example</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/argocd/.env.example

<ul><li>Updated environment variable format to use OCEAN__ prefix<br> <li> Added missing configuration variables for integration setup<br> <li> Fixed syntax formatting for environment variables</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/1923/files#diff-5e07e76794162cad3e12563d3f03a86c082fd2a5e115a07ea8a9e6bd4007d71d">+8/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Review by mk-armah — CHANGES_REQUESTED** [review]
Add changelog and bump integration version

File: integrations/argocd/client.py [hunk 1]
```diff
@@ -100,10 +100,13 @@ async def get_deployment_history(self) -> list[dict[str, Any]]:
             f"get_deployment_history is deprecated as of 0.1.34. {DEPRECATION_WARNING}"
         )
         applications = await self.get_resources(resource_kind=ObjectKind.APPLICATION)
+        if applications is None:
+            return []
         all_history = [
             {**history_item, "__applicationId": application["metadata"]["uid"]}
             for application in applications
-            for history_item in application["status"].get("history", [])
+            if application and application.get("metadata", {}).get("uid")
```
Comment (mk-armah) [inline integrations/argocd/client.py]:
```suggestion
            if application
```

File: integrations/argocd/client.py [hunk 2]
```diff
@@ -113,17 +116,52 @@ async def get_kubernetes_resource(self) -> list[dict[str, Any]]:
...
+            logger.info(
+                f"Fetching managed resources for application: {application_name}"
+            )
+            url = f"{self.api_url}/{ObjectKind.APPLICATION}s/{application_name}/managed-resources"
+            managed_resources = (await self._send_api_request(url=url)).get("items", [])
+            application_resource = [
+                {
+                    **managed_resource,
+                    "__application": application,
+                    "__applicationId": application_id,
```
Comment (mk-armah) [inline integrations/argocd/client.py]: investigate from NR if any `__applicationId` is being used by any client, if not, remove

File: integrations/argocd/client.py [hunk 3]
```diff
@@ -113,17 +116,52 @@ async def get_kubernetes_resource(self) -> list[dict[str, Any]]:
...
+    ) -> list[dict[str, Any]] | None:
+        errors = []
+        try:
+            application_name = application.get("metadata", {}).get("name")
+            application_id = application.get("metadata", {}).get("uid")
+            if not application_name or not application_id:
+                logger.error(
+                    "Application metadata is missing 'name' or 'uid', skipping..."
+                )
+                return []
```
Comment (mk-armah) [inline integrations/argocd/client.py]: Is there a possibility to receive an application object with missing name and uid fields ?

Comment (dennis-quabynah-bilson) [reply integrations/argocd/client.py]: @mk-armah Very unlikely but just in case it arises, we won't miss it

File: integrations/argocd/client.py [hunk 4]
```diff
@@ -100,10 +100,13 @@ async def get_deployment_history(self) -> list[dict[str, Any]]:
             f"get_deployment_history is deprecated as of 0.1.34. {DEPRECATION_WARNING}"
         )
         applications = await self.get_resources(resource_kind=ObjectKind.APPLICATION)
+        if applications is None:
+            return []
```
Comment (nivm-port) [inline integrations/argocd/client.py:106]: add log before return

File: integrations/argocd/client.py [hunk 5]
```diff
@@ -113,17 +116,52 @@ async def get_kubernetes_resource(self) -> list[dict[str, Any]]:
             f"get_kubernetes_resource is deprecated as of 0.1.34. {DEPRECATION_WARNING}"
         )
         applications = await self.get_resources(resource_kind=ObjectKind.APPLICATION)
+        if applications is None:
+            return []
```
Comment (nivm-port) [inline integrations/argocd/client.py:125]: add log before return

File: integrations/argocd/main.py [hunk 6]
```diff
@@ -46,36 +47,15 @@ async def on_managed_resources_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
...
-            errors.append(e)
-
-    # If there were errors and we are not ignoring server errors, raise a general exception
-    if errors and not argocd_client.ignore_server_error:
-        raise ExceptionGroup(
-            "Errors occurred during managed resource ingestion", errors
+        managed_resources = await argocd_client.get_managed_resources(
+            application=application
         )
+        if managed_resources:
```
Comment (nivm-port) [inline integrations/argocd/main.py]: add log

**Review by nivm-port — CHANGES_REQUESTED** [review]
Add some logs but except this good job :)

File: integrations/argocd/main.py [hunk 7]
```diff
@@ -46,36 +45,18 @@ async def on_managed_resources_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
...
-
-    # If there were errors and we are not ignoring server errors, raise a general exception
-    if errors and not argocd_client.ignore_server_error:
-        raise ExceptionGroup(
-            "Errors occurred during managed resource ingestion", errors
+        managed_resources = await argocd_client.get_managed_resources(
+            application=application
+        )
+        logger.info(
+            f"Ingesting managed resources for application: {application.get('metadata', {}).get('name')}"
```
Comment (mk-armah) [inline integrations/argocd/main.py]:
```suggestion
      if application:
          managed_resources = await argocd_client.get_managed_resources(
              application=application
          )
          logger.info(
              f"Ingesting managed resources for application: {application.get('metadata', {}).get('name')}"
```

File: integrations/argocd/client.py [hunk 8]
```diff
@@ -113,17 +119,55 @@ async def get_kubernetes_resource(self) -> list[dict[str, Any]]:
...
-        url = f"{self.api_url}/{ObjectKind.APPLICATION}s/{application_name}/managed-resources"
-        managed_resources = (await self._send_api_request(url=url)).get("items", [])
-        return managed_resources
+        self, application: dict[str, Any]
+    ) -> list[dict[str, Any]] | None:
+        errors = []
+        try:
+            application_name = application.get("metadata", {}).get("name")
+            application_id = application.get("metadata", {}).get("uid")
+            if not application_name or not application_id:
```
Comment (mk-armah) [inline integrations/argocd/client.py]:
```suggestion
```

Comment (mk-armah) [reply integrations/argocd/client.py]: Verify whether or not metadata is guaranteed

**Review by mk-armah — CHANGES_REQUESTED** [review]
.__applicationId

File: integrations/argocd/client.py (hunk 8 above)
Comment (dennis-quabynah-bilson) [reply integrations/argocd/client.py]: It is guaranteed
https://cd.apps.argoproj.io/swagger-ui#tag/ApplicationService
